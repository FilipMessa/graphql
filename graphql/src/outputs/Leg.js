// @flow

import { GraphQLObjectType, GraphQLBoolean, GraphQLInt } from 'graphql';
import { globalIdField } from '../services/OpaqueIdentifier';

import GraphQLRouteStop from './RouteStop';
import GraphQLAirline from './Airline';
import FlightDurationInMinutes from '../resolvers/FlightDuration';

import type { GraphqlContextType } from '../services/GraphqlContext';
import type { DepartureArrival, Leg } from '../types/Flight';

export default new GraphQLObjectType({
  name: 'Leg',
  description:
    'Leg is the operation of an aircraft from one scheduled departure station to its next scheduled arrival station.',
  fields: {
    id: globalIdField(),

    airline: {
      type: GraphQLAirline,
      resolve: async (
        { airlineCode }: Leg,
        args: Object,
        { dataLoader }: GraphqlContextType,
      ) => dataLoader.airline.load(airlineCode),
    },

    arrival: {
      type: GraphQLRouteStop,
      resolve: ({ arrival }: Leg): DepartureArrival => arrival,
    },

    departure: {
      type: GraphQLRouteStop,
      resolve: ({ departure }: Leg): DepartureArrival => departure,
    },

    duration: {
      type: GraphQLInt,
      description: 'Leg duration in minutes.',
      resolve: ({ departure, arrival }: Leg): ?number =>
        FlightDurationInMinutes(departure, arrival),
    },

    flightNumber: {
      type: GraphQLInt,
      resolve: ({ flightNo }: Leg): number => flightNo,
    },

    recheckRequired: {
      type: GraphQLBoolean,
      resolve: ({ recheckRequired }): boolean => recheckRequired,
    },
  },
});
