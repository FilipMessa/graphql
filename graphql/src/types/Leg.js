// @flow

import {
  GraphQLObjectType,
  GraphQLNonNull,
  GraphQLID,
  GraphQLBoolean,
  GraphQLInt,
} from 'graphql';
import { toGlobalId } from '../services/OpaqueIdentifier';
import GraphQLRouteStop from './RouteStop';
import GraphQLAirline from './Airline';
import { createAirline } from '../dataLoaders/Airline';
import { flightDurationInMinutes } from '../services/GraphqlResolvers';

import type {
  ArrivalType,
  DepartureType,
  LegType,
  AirlineType,
} from '../Entities';

export default new GraphQLObjectType({
  name: 'Leg',
  description: 'Leg is the operation of an aircraft from one scheduled departure station to its next scheduled arrival station.',
  fields: {
    id: {
      type: new GraphQLNonNull(GraphQLID),
      resolve: ({ id }: LegType): string => toGlobalId('leg', id),
    },

    airline: {
      type: GraphQLAirline,
      resolve: ({ airline }: LegType): AirlineType => createAirline(airline),
    },

    arrival: {
      type: new GraphQLNonNull(GraphQLRouteStop),
      resolve: ({ arrival }: LegType): ArrivalType => arrival,
    },

    departure: {
      type: new GraphQLNonNull(GraphQLRouteStop),
      resolve: ({ departure }: LegType): DepartureType => departure,
    },

    duration: {
      type: GraphQLInt,
      resolve: ({ departure, arrival }: LegType): ?number =>
        flightDurationInMinutes(departure, arrival),
    },

    flightNumber: {
      type: GraphQLInt,
      resolve: ({ flightNo }: LegType): number => flightNo,
    },

    recheckRequired: {
      type: new GraphQLNonNull(GraphQLBoolean),
      resolve: ({ recheckRequired }): boolean => recheckRequired,
    },
  },
});
