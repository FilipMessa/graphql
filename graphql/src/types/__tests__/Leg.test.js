// @flow

import Leg from '../Leg';

const fields = Leg.getFields();

it('Leg type should have description', () => {
  expect(Leg.description).not.toBe(undefined);
});

it('Field "id" should be non-null ID type', () => {
  expect(fields.id.type.toString()).toBe('ID!');
});

it('Field "id" should use opaque identifiers', () => {
  expect(fields.id.resolve({ id: 1 })).toBe('bGVnOjE='); // leg:1
});

it('Field "arrival" should be non-null Arrival type', () => {
  expect(fields.arrival.type.toString()).toBe('Arrival!');
});

it('Field "Departure" should be non-null ID type', () => {
  expect(fields.departure.type.toString()).toBe('Departure!');
});

it('Field "recheckRequired" should be non-null Boolean type', () => {
  expect(fields.recheckRequired.type.toString()).toBe('Boolean!');
});
