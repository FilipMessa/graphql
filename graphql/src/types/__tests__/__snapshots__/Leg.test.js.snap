// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Leg type should have valid fields 1`] = `
Object {
  "airline": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "airline",
    "resolve": [Function],
    "type": "Airline",
  },
  "arrival": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "arrival",
    "resolve": [Function],
    "type": "RouteStop!",
  },
  "departure": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "departure",
    "resolve": [Function],
    "type": "RouteStop!",
  },
  "duration": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "duration",
    "resolve": [Function],
    "type": "Int",
  },
  "flightNumber": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "flightNumber",
    "resolve": [Function],
    "type": "Int",
  },
  "id": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "id",
    "resolve": [Function],
    "type": "ID!",
  },
  "recheckRequired": Object {
    "args": Array [],
    "isDeprecated": false,
    "name": "recheckRequired",
    "resolve": [Function],
    "type": "Boolean!",
  },
}
`;
